File: Data Structures & Algorithms/valid-parenthesis-string/submission-2.cpp
class Solution {
private:
    unordered_map<int, unordered_map<int, bool> > memo;
    bool dfs(string s, int sz, int idx) {
        if (idx >= s.size()) return sz == 0;

        char curr = s[idx];

        if (memo.contains(sz) && memo[sz].contains(idx)) return memo[sz][idx];

        bool res = false;
        switch (curr) {
            case '(':
                res |= dfs(s, sz + 1, idx + 1);
                break;
            case ')':
                if (sz <= 0) return false;
                res |= dfs(s, sz - 1, idx + 1);
                break;
            case '*':
                res |= dfs(s, sz + 1, idx + 1);
                
                if (sz > 0) {
                    res |= dfs(s, sz - 1, idx + 1);
                }
                
                res |= dfs(s, sz, idx + 1);
        }

        return memo[sz][idx] = res;
    }

public:
    bool checkValidString(string s) {
        return dfs(s, 0, 0);
    }
};

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-2.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        children = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        def dfs(idx, curr):
            return res.append(curr) if idx == len(digits) else [dfs(idx + 1, curr + c) for c in children[digits[idx]]]

        return res if digits and (dfs(0, "") or 1) else []
